src/oidcendpoint/session/database.py

```diff
@@ -0,0 +1,173 @@
+
+        _userinfo = self._db.get(uid)
+        if _userinfo:
+            if client_id:
+                if client_id in _userinfo['subordinate']:
+                    _cid_key = session_key(uid, client_id)
+                    _cid_info = self._db[session_key(uid, client_id)]
+                    # if _cid_info.is_revoked():
+                    #     raise Revoked("Session is revoked")
+                    if _cid_info:
+                        if grant_id:
+                            _gid_key = session_key(uid, client_id, grant_id)
+                            if grant_id in _cid_info['subordinate']:
+                                _gid_info = self._db[_gid_key]
+                                if not _gid_info:
+                                    self._db[_cid_key] = _cid_info.add_subordinate(grant_id)
+                                self._db[_gid_key] = value
+                            else:
+                                self._db[_cid_key] = _cid_info.add_subordinate(grant_id)
+                                self._db[_gid_key] = value
+                        else:
+                            self._db[_cid_key] = value
+                    else:
+                        _userinfo.add_subordinate(client_id)
+                        if grant_id:
+                            _cid_info = ClientSessionInfo()
+                            _cid_info.add_subordinate(grant_id)
+                            self._db[_cid_key] = _cid_info
+                            self._db[session_key(uid, client_id, grant_id)] = value
+                        else:
+                            _cid_info = ClientSessionInfo()
+                            self._db[_cid_key] = _cid_info
+                        self._db[uid] = _userinfo
+                else:
+                    _userinfo.add_subordinate(client_id)
+                    self._db[uid] = _userinfo
+                    if grant_id:
+                        _cid_info = ClientSessionInfo()
+                        _cid_info.add_subordinate(grant_id)
+                        self._db[session_key(uid, client_id, grant_id)] = value
+                    else:
+                        _cid_info = value
+
+                    _cid_key = session_key(uid, client_id)
+                    self._db[_cid_key] = _cid_info
+            else:
+                self._db[uid] = value
+        else:
+            if client_id:
+                _user_info = UserSessionInfo()
+                _user_info.add_subordinate(client_id)
+                if grant_id:
+                    _cid_info = ClientSessionInfo()
+                    _cid_info.add_subordinate(grant_id)
+                    self._db[session_key(uid, client_id, grant_id)] = value
+                else:
+                    _cid_info = value
+                self._db[session_key(uid, client_id)] = _cid_info
+            else:
+                _user_info = value
+
+            self._db[uid] = _user_info
```

Comment:
I think this can be refactored to:
```python
        uid, client_id, grant_id = self._eval_path(path)

        if grant_id:
            gid_key = session_key(uid, client_id, grant_id)
            self._db[gid_key] = value

        if client_id:
            cid_key = session_key(uid, client_id)
            if not grant_id:
                self._db[cid_key] = value
            elif grant_id not in cid_info["subordinate"]:
                cid_info = self._db.get(cid_key, ClientSessionInfo())
                cid_info.add_subordinate(grant_id)
                self._db[cid_key] = cid_info

        userinfo = self._db.get(uid, UserSessionInfo())
        if not client_id:
            self._db[uid] = value
        if client_id not in userinfo["subordinate"]:
            userinfo.add_subordinate(client_id)
            self._db[uid] = userinfo
```

Reply:
Nope!
Tested with the extended set of tests and a bunch failed.
Still it might be worth seeing if your refactoring only needs a small set of tweaks to work.

Reply:
Just needed a few modifications.